src/card_operations/card_internal.c

```diff
@@ -0,0 +1,368 @@
+    NFC_RETURN_ABORT_ERROR(card_data, ui_text_unknown_error_contact_support);
+  } else if ((card_data->nfc_data.status & NFC_ERROR_BASE) == NFC_ERROR_BASE) {
+    instruction_scr_change_text(ui_text_card_align_with_device_screen, true);
+    nfc_deselect_card();
+  }
+
+  // probably comm failure; guess is NFC teardown; reconnect with applet
+  // silently retry to connect; can't connect here, cards_state is important
+  NFC_RETURN_SUCCESS(card_data);
+}
+
+static card_error_type_e handle_wait_for_card_selection(
+    card_operation_data_t *card_data) {
+  evt_status_t evt_status = {0};
+
+  while (1) {
+    nfc_en_select_card_task();
+
+    evt_status = get_events(EVENT_CONFIG_NFC, MAX_INACTIVITY_TIMEOUT);
+
+    HANDLE_P0_EVENTS(evt_status.p0_event);
+
+    if (true == evt_status.nfc_event.event_occured) {
+      if (STM_SUCCESS == nfc_wait_for_card(DEFAULT_NFC_TG_INIT_TIME))
+        NFC_RETURN_SUCCESS(card_data);
+    }
+  }
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+card_error_type_e card_initialize_applet(card_operation_data_t *card_data) {
+  card_data->error_type = CARD_OPERATION_DEFAULT_INVALID;
+  card_data->nfc_data.recovery_mode = 0;
+  if (!card_data->nfc_data.card_absent_retries) {
+    card_data->nfc_data.card_absent_retries = 100;
+  }
+
+  while (1) {
+    if (CARD_OPERATION_SUCCESS != handle_wait_for_card_selection(card_data)) {
+      return card_data->error_type;
+    }
+
+    instruction_scr_change_text(ui_text_card_detected, true);
+
+    select_applet_and_update_tapped_card(&(card_data->nfc_data));
+
+    /* Card is in recovery mode. This is a critical situation. Instruct user to
+     * safely recover/export wallets to different set of cards in limited
+     * attempts. NOTE: Errors such as invalid card & invalid family id have
+     * higher priority than this.
+     */
+    if (card_data->nfc_data.recovery_mode == 1) {
+      NFC_RETURN_ABORT_ERROR(card_data, ui_critical_card_health_migrate_data);
+    }
+
+    LOG_ERROR("Applet selection err (0x%04X)\n", card_data->nfc_data.status);
+    switch (card_data->nfc_data.status) {
+      case SW_NO_ERROR:
+        return load_session_key_if_pairing_required(card_data);
+        break;
+      case SW_CONDITIONS_NOT_SATISFIED:
+        NFC_RETURN_RETAP_ERROR(card_data, ui_text_wrong_card_sequence);
+        break;
+      case SW_FILE_INVALID:
+        NFC_RETURN_RETAP_ERROR(card_data, ui_text_family_id_mismatch);
+        break;
+      case SW_FILE_NOT_FOUND:
+        NFC_RETURN_ABORT_ERROR(card_data,
+                               ui_text_corrupted_card_contact_support);
+        break;
+      case SW_INCOMPATIBLE_APPLET:
+        NFC_RETURN_ABORT_ERROR(card_data, ui_text_incompatible_card_version);
+        break;
+      default:
+        card_data->nfc_data.acceptable_cards |= card_data->nfc_data.tapped_card;
```

Comment:
`tap_card_data.tapped_card = 0;` has been removed. could have unexpected impact.

Reply:
Not removed but moved to the function call in next line.